File: src/edu/uf/intracellularState/AspergillusIntracellularModel.java
package edu.uf.intracellularState;

import edu.uf.interactable.Cell;
import edu.uf.interactable.Afumigatus.Afumigatus;
import edu.uf.utils.Constants;
import edu.uf.utils.Rand;
import edu.uf.utils.Util;

public class AspergillusIntracellularModel extends IntracellularModel{
	
	public static final int[] INIT_AFUMIGATUS_BOOLEAN_STATE = new int[] {1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    
    private static final int hapX = 0;
    private static final int sreA = 1;
    private static final int HapX = 2;
    private static final int SreA = 3;
    private static final int RIA = 4;
    private static final int EstB = 5;
    private static final int MirB = 6;
    private static final int SidA = 7;
    private static final int Tafc = 8;
    private static final int ICP = 9;
    private static final int LIP = 10;
    private static final int CccA = 11;
    private static final int FC0fe = 12;
    private static final int FC1fe = 13;
    private static final int VAC = 14;
    private static final int ROS = 15;
    private static final int Yap1 = 16;
    private static final int SOD2_3 = 17;
    private static final int Cat1_2 = 18;
    private static final int ThP = 19;
    private static final int Fe = 20;
    private static final int O = 21;
    private static final int NUM_SPECIES = 22;
    
    
    
    public static final int SECRETING_TAFC = Phenotype.createPhenotype();
    public static final int UPTAKING_TAFC = Phenotype.createPhenotype();
    
    //public static final int Af_STATUS_START = RESTING_CONIDIA;
    
    private int lipActivation;
    
    public AspergillusIntracellularModel(final int phenotype) {
    	setBnIteration(0);
    	setBooleanNetwork(INIT_AFUMIGATUS_BOOLEAN_STATE.clone());
    	//System.out.println(LOCATION + " " + STATUS);
    	super.setState(LOCATION, Afumigatus.FREE);
    	super.setState(STATUS, phenotype);
    	//super.addPhenotype(INTERACTION_STATE, phenotype);
    	
    }

    static int i = 0;
    static int j = 0;
	@Override
	public void updateStatus(int id, int x, int y, int z) {
		Afumigatus asp = (Afumigatus) Cell.get(id);
		
		this.lipActivation = Rand.getRand().randunif() < Util.activationFunction(asp.getIronPool(), Constants.Kd_LIP, Constants.HYPHAE_VOL, 1.0) ? 1 : 0;
		//System.out.println(lipActivation);
		
		//System.out.println(this.hasPhenotype(RESTING_CONIDIA));
		//System.out.println(Constants.PR_ASPERGILLUS_CHANGE);
		//System.out.println(Rand.getRand().randunif());
		if(this.getState(STATUS) == Afumigatus.RESTING_CONIDIA && asp.getClock().getCount() >= Constants.ITER_TO_SWELLING)j++;
        if(this.getState(STATUS) == Afumigatus.RESTING_CONIDIA && asp.getClock().getCount() >= Constants.ITER_TO_SWELLING && Rand.getRand().randunif() < Constants.PR_ASPERGILLUS_CHANGE) { 
            this.setState(STATUS, Afumigatus.SWELLING_CONIDIA);
            Afumigatus.incTotalCells(2);
			Afumigatus.decTotalCells(1);
			i++;
			//System.out.println(i/((double)j));
        }else if(!asp.isEngaged() && this.getState(STATUS) == Afumigatus.SWELLING_CONIDIA && asp.getClock().getCount() >= Constants.ITER_TO_GERMINATE) {
            this.setState(STATUS, Afumigatus.GERM_TUBE);
            Afumigatus.incTotalCells(3);
			Afumigatus.decTotalCells(2);
        }else if(this.getState(IntracellularModel.LIFE_STATUS) == IntracellularModel.DYING) {
        	asp.die();
        }

        if(asp.getNextSepta() == null) 
        	asp.setGrowable(true);

        if(this.getState(LOCATION) == Afumigatus.INTERNALIZING || this.getState(LOCATION) == Afumigatus.RELEASING)
        	this.setState(LOCATION, Afumigatus.FREE); 

        asp.diffuseIron();
        if(asp.getNextBranch() == null)
        	asp.setGrowable(true);
        asp.setEpithelialInhibition(2.0);
		
	}
	
	public static int hasLifeStage(int phenotype) {
		if(phenotype == Afumigatus.RESTING_CONIDIA)
			return 1;
		if(phenotype == Afumigatus.SWELLING_CONIDIA)
			return 2;
		if(phenotype == Afumigatus.GERM_TUBE)
			return 3;
		if(phenotype == Afumigatus.HYPHAE)
			return 4;
		if(phenotype == Afumigatus.STERILE_CONIDIA) //PROBABLY NOT BEEN CONSIDERED
			return 5; //THAT'S TOO MUCH. PROBABLY GOING TO RAISE AN ERROR
		return -1;
	}

	@Override
	public void processBooleanNetwork(int... args) {
		int[] temp = new int[NUM_SPECIES];
		
        temp[hapX] = -this.booleanNetwork[SreA] + 1;
        temp[sreA] = -this.booleanNetwork[HapX] + 1;
        temp[HapX] = this.booleanNetwork[hapX] & (-this.booleanNetwork[LIP] + 1);
        temp[SreA] = this.booleanNetwork[sreA] & this.booleanNetwork[LIP];
        temp[RIA] = -this.booleanNetwork[SreA] + 1;
        temp[EstB] = -this.booleanNetwork[SreA] + 1;
        temp[MirB] = this.booleanNetwork[HapX] & (-this.booleanNetwork[SreA] + 1);
        temp[SidA] = this.booleanNetwork[HapX] & (-this.booleanNetwork[SreA] + 1);
        temp[Tafc] = this.booleanNetwork[SidA];
        temp[ICP] = (-this.booleanNetwork[HapX] + 1) & (this.booleanNetwork[VAC] | this.booleanNetwork[FC1fe]);
        temp[LIP] = (this.booleanNetwork[Fe] & this.booleanNetwork[RIA]) | this.lipActivation;
        temp[CccA] = -this.booleanNetwork[HapX] + 1;
        temp[FC0fe] = this.booleanNetwork[SidA];
        temp[FC1fe] = this.booleanNetwork[LIP] & this.booleanNetwork[FC0fe];
        temp[VAC] = this.booleanNetwork[LIP] & this.booleanNetwork[CccA];
        temp[ROS] = (this.booleanNetwork[O] & (- (this.booleanNetwork[SOD2_3] & this.booleanNetwork[ThP] 
                               & this.booleanNetwork[Cat1_2]) + 1)) 
                              | (this.booleanNetwork[ROS] & (- (this.booleanNetwork[SOD2_3] 
                               & (this.booleanNetwork[ThP] | this.booleanNetwork[Cat1_2])) + 1));
        temp[Yap1] = this.booleanNetwork[ROS];
        temp[SOD2_3] = this.booleanNetwork[Yap1];
        temp[Cat1_2] = this.booleanNetwork[Yap1] & (-this.booleanNetwork[HapX] + 1);
        temp[ThP] = this.booleanNetwork[Yap1];

        temp[Fe] = 0; // might change according to iron environment?
        temp[O] = 0;
        //temp[Afumigatus.TAFCBI] = 0;

        //print(this.boolean_network)
        for(int i = 0; i < NUM_SPECIES; i++)
            this.booleanNetwork[i] = temp[i];
        this.setBnIteration(0);
        
        this.computePhenotype();
		
	}

	@Override
	protected void computePhenotype() {
		if(this.booleanNetwork[MirB] == 1 && this.booleanNetwork[EstB] == 1)
			this.addPhenotype(UPTAKING_TAFC);
		if(this.booleanNetwork[Tafc] == 1)
			this.addPhenotype(SECRETING_TAFC);
		
	}

}

File: cpp/src/edu/uf/intracellularState/AspergillusIntracellularModel.cpp
/*
 * AspergillusIntracellularModel.cpp
 *
 *  Created on: Jul 18, 2024
 *      Author: henriquedeassis
 */

#include "AspergillusIntracellularModel.h"
#include "../interactable/afumigatus/Afumigatus.h"
#include "../interactable/Cell.h"
#include "../utils/Constants.h"
#include "../utils/Rand.h"
#include "../utils/Util.h"
#include "../utils/Id.h"

namespace edu {
namespace uf {
namespace intracellularState {

using namespace edu::uf::utils;
using namespace afumigatus;

int AspergillusIntracellularModel::INIT_AFUMIGATUS_BOOLEAN_STATE[] = {1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};


const int AspergillusIntracellularModel::RESTING_CONIDIA = Id::getId();
const int AspergillusIntracellularModel::SWELLING_CONIDIA = Id::getId();
const int AspergillusIntracellularModel::GERM_TUBE = Id::getId();
const int AspergillusIntracellularModel::HYPHAE = Id::getId();
const int AspergillusIntracellularModel::STERILE_CONIDIA = Id::getId();

const int AspergillusIntracellularModel::FREE = Id::getId();
const int AspergillusIntracellularModel::INTERNALIZING = Id::getId();
const int AspergillusIntracellularModel::RELEASING = Id::getId();
const int AspergillusIntracellularModel::ENGAGED = Id::getId();

const int AspergillusIntracellularModel::SECRETING_TAFC = Id::getId();
const int AspergillusIntracellularModel::UPTAKING_TAFC = Id::getId();

const int AspergillusIntracellularModel::NUM_SPECIES = 22;

AspergillusIntracellularModel::AspergillusIntracellularModel(int phenotype) : IntracellularModel() {
    setBnIteration(0);
    setBooleanNetwork(&INIT_AFUMIGATUS_BOOLEAN_STATE[0]);
    setState(LOCATION, AspergillusIntracellularModel::FREE);
    setState(STATUS, phenotype);
    this->lipActivation = 0;
}

void AspergillusIntracellularModel::updateStatus(int id, int x, int y, int z) {
    Afumigatus* asp = dynamic_cast<Afumigatus*>(Cell::get(id));

    lipActivation = Rand::getRand().randunif() < Util::activationFunction(asp->getIronPool(), utils::constexprants::Kd_LIP, utils::constexprants::HYPHAE_VOL) ? 1 : 0;

    if (getState(STATUS) == AspergillusIntracellularModel::RESTING_CONIDIA && asp->getClock()->getCount() >= utils::constexprants::ITER_TO_SWELLING) {
        if (Rand::getRand().randunif() < utils::constexprants::PR_ASPERGILLUS_CHANGE) {
            setState(STATUS, AspergillusIntracellularModel::SWELLING_CONIDIA);
            Afumigatus::incTotalCells(2);
            Afumigatus::decTotalCells(1);
        }
    } else if (!asp->isEngaged() && getState(STATUS) == AspergillusIntracellularModel::SWELLING_CONIDIA && asp->getClock()->getCount() >= utils::constexprants::ITER_TO_GERMINATE) {
        setState(STATUS, AspergillusIntracellularModel::GERM_TUBE);
        Afumigatus::incTotalCells(3);
        Afumigatus::decTotalCells(2);
    } else if (getState(IntracellularModel::LIFE_STATUS) == IntracellularModel::DYING) {
        asp->die();
    }

    if (asp->getNextSepta() == nullptr)
        asp->setGrowable(true);

    if (getState(LOCATION) == AspergillusIntracellularModel::INTERNALIZING || getState(LOCATION) == AspergillusIntracellularModel::RELEASING)
        setState(LOCATION, AspergillusIntracellularModel::FREE);

    asp->diffuseIron();
    if (asp->getNextBranch() == nullptr)
        asp->setGrowable(true);
    asp->setEpithelialInhibition(2.0);
}

int AspergillusIntracellularModel::hasLifeStage(int phenotype) {
    if (phenotype == AspergillusIntracellularModel::RESTING_CONIDIA)
        return 1;
    if (phenotype == AspergillusIntracellularModel::SWELLING_CONIDIA)
        return 2;
    if (phenotype == AspergillusIntracellularModel::GERM_TUBE)
        return 3;
    if (phenotype == AspergillusIntracellularModel::HYPHAE)
        return 4;
    if (phenotype == AspergillusIntracellularModel::STERILE_CONIDIA)
        return 5;
    return -1;
}

void AspergillusIntracellularModel::processBooleanNetwork() {
    int temp[NUM_SPECIES] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

    temp[hapX] = -booleanNetwork[sreA] + 1;
    temp[sreA] = -booleanNetwork[HapX] + 1;
    temp[HapX] = booleanNetwork[hapX] & (-booleanNetwork[LIP] + 1);
    temp[SreA] = booleanNetwork[sreA] & booleanNetwork[LIP];
    temp[RIA] = -booleanNetwork[SreA] + 1;
    temp[EstB] = -booleanNetwork[SreA] + 1;
    temp[MirB] = booleanNetwork[HapX] & (-booleanNetwork[SreA] + 1);
    temp[SidA] = booleanNetwork[HapX] & (-booleanNetwork[SreA] + 1);
    temp[Tafc] = booleanNetwork[SidA];
    temp[ICP] = (-booleanNetwork[HapX] + 1) & (booleanNetwork[VAC] | booleanNetwork[FC1fe]);
    temp[LIP] = (booleanNetwork[Fe] & booleanNetwork[RIA]) | lipActivation;
    temp[CccA] = -booleanNetwork[HapX] + 1;
    temp[FC0fe] = booleanNetwork[SidA];
    temp[FC1fe] = booleanNetwork[LIP] & booleanNetwork[FC0fe];
    temp[VAC] = booleanNetwork[LIP] & booleanNetwork[CccA];
    temp[ROS] = (booleanNetwork[O] & (-(booleanNetwork[SOD2_3] & booleanNetwork[ThP] & booleanNetwork[Cat1_2]) + 1))
                | (booleanNetwork[ROS] & (-(booleanNetwork[SOD2_3] & (booleanNetwork[ThP] | booleanNetwork[Cat1_2])) + 1));
    temp[Yap1] = booleanNetwork[ROS];
    temp[SOD2_3] = booleanNetwork[Yap1];
    temp[Cat1_2] = booleanNetwork[Yap1] & (-booleanNetwork[HapX] + 1);
    temp[ThP] = booleanNetwork[Yap1];

    temp[Fe] = 0;
    temp[O] = 0;

    for (int i = 0; i < NUM_SPECIES; ++i) {
        booleanNetwork[i] = temp[i];
    }
    setBnIteration(0);
    computePhenotype();
}

void AspergillusIntracellularModel::computePhenotype() {
    if (booleanNetwork[MirB] == 1 && booleanNetwork[EstB] == 1)
        addPhenotype(UPTAKING_TAFC);
    if (booleanNetwork[Tafc] == 1)
        addPhenotype(SECRETING_TAFC);
}

} // namespace intracellularState
} // namespace uf
} // namespace edu
